autogpt_platform/frontend/src/components/integrations/credentials-input.tsx

```diff
@@ -95,56 +96,99 @@ export const CredentialsInput: FC<{
+
+      const handleMessage = async (
+        e: MessageEvent<OAuthPopupResultMessage>,
+      ) => {
+        console.debug("Message received:", e.data);
         if (
           typeof e.data != "object" ||
-          !(
-            "message_type" in e.data &&
-            e.data.message_type == "oauth_popup_result"
-          )
-        )
+          !("message_type" in e.data) ||
+          e.data.message_type !== "oauth_popup_result"
+        ) {
+          console.debug("Ignoring irrelevant message");
           return;
+        }
 
         if (!e.data.success) {
           console.error("OAuth flow failed:", e.data.message);
+          setOAuthError(`OAuth flow failed: ${e.data.message}`);
+          setOAuth2FlowInProgress(false);
+          return;
+        }
+
+        if (e.data.state !== state_token) {
+          console.error("Invalid state token received");
+          setOAuthError("Invalid state token received");
+          setOAuth2FlowInProgress(false);
           return;
         }
 
-        if (e.data.state !== state_token) return;
+        try {
+          console.debug("Processing OAuth callback");
+          const credentials = await oAuthCallback(e.data.code, e.data.state);
+          console.debug("OAuth callback processed successfully");
+          onSelectCredentials({
+            id: credentials.id,
+            type: "oauth2",
+            title: credentials.title,
+            provider,
+          });
+        } catch (error) {
+          console.error("Error in OAuth callback:", error);
+          setOAuthError(
+            `Error in OAuth callback: ${error instanceof Error ? error.message : String(error)}`,
+          );
+        } finally {
+          console.debug("Finalizing OAuth flow");
+          setOAuth2FlowInProgress(false);
+          controller.abort("success");
+        }
+      };
 
-        const credentials = await oAuthCallback(e.data.code, e.data.state);
-        onSelectCredentials({
-          id: credentials.id,
-          type: "oauth2",
-          title: credentials.title,
-          provider,
-        });
-        controller.abort("success");
-      },
-      { signal: controller.signal },
-    );
+      console.debug("Adding message event listener");
+      window.addEventListener("message", handleMessage, {
+        signal: controller.signal,
+      });
 
-    setTimeout(
-      () => {
-        controller.abort("timeout");
-      },
-      5 * 60 * 1000,
-    );
+      setTimeout(
+        () => {
+          console.debug("OAuth flow timed out");
+          controller.abort("timeout");
+          setOAuth2FlowInProgress(false);
+          setOAuthError("OAuth flow timed out");
+        },
+        5 * 60 * 1000,
+      );
+    } catch (error) {
+      console.error("Error initiating OAuth flow:", error);
+      setOAuthError(
+        `Error initiating OAuth flow: ${error instanceof Error ? error.message : String(error)}`,
+      );
+      setOAuth2FlowInProgress(false);
+    }
```

Comment:
Any chance you could make this try/catch more specific? Now there are multiple nested try/catch blocks which doesn't make sense to me.

Reply:
shouldn't have been nested. OOPS